# Customer Churn Prediction using Machine Learning


In [2]:
import pandas as pd

train = pd.read_csv('customer_churn_dataset-training-master.csv')
test = pd.read_csv('customer_churn_dataset-testing-master.csv')

train.head()

,CustomerID,Age,Gender,Tenure,Usage Frequency,Support Calls,Payment Delay,Subscription Type,Contract Length,Total Spend,Last Interaction,Churn
0,2.0,30.0,Female,39.0,14.0,5.0,18.0,Standard,Annual,932.0,17.0,1.0
1,3.0,65.0,Female,49.0,1.0,10.0,8.0,Basic,Monthly,557.0,6.0,1.0
2,4.0,55.0,Female,14.0,4.0,6.0,18.0,Basic,Quarterly,185.0,3.0,1.0
3,5.0,58.0,Male,38.0,21.0,7.0,7.0,Standard,Monthly,396.0,29.0,1.0
4,6.0,23.0,Male,32.0,20.0,5.0,8.0,Basic,Monthly,617.0,20.0,1.0


In [3]:
train.columns

Index(['CustomerID', 'Age', 'Gender', 'Tenure', 'Usage Frequency',
       'Support Calls', 'Payment Delay', 'Subscription Type',
       'Contract Length', 'Total Spend', 'Last Interaction', 'Churn'],
      dtype='object')

In [4]:
from sklearn.model_selection import train_test_split

# features (all except from churn)
X = train.drop('Churn', axis=1)

# target (the one we predict)
y = train['Churn']

X.head()

,CustomerID,Age,Gender,Tenure,Usage Frequency,Support Calls,Payment Delay,Subscription Type,Contract Length,Total Spend,Last Interaction
0,2.0,30.0,Female,39.0,14.0,5.0,18.0,Standard,Annual,932.0,17.0
1,3.0,65.0,Female,49.0,1.0,10.0,8.0,Basic,Monthly,557.0,6.0
2,4.0,55.0,Female,14.0,4.0,6.0,18.0,Basic,Quarterly,185.0,3.0
3,5.0,58.0,Male,38.0,21.0,7.0,7.0,Standard,Monthly,396.0,29.0
4,6.0,23.0,Male,32.0,20.0,5.0,8.0,Basic,Monthly,617.0,20.0


In [5]:
X = X.drop('CustomerID', axis=1)
X.head()

,Age,Gender,Tenure,Usage Frequency,Support Calls,Payment Delay,Subscription Type,Contract Length,Total Spend,Last Interaction
0,30.0,Female,39.0,14.0,5.0,18.0,Standard,Annual,932.0,17.0
1,65.0,Female,49.0,1.0,10.0,8.0,Basic,Monthly,557.0,6.0
2,55.0,Female,14.0,4.0,6.0,18.0,Basic,Quarterly,185.0,3.0
3,58.0,Male,38.0,21.0,7.0,7.0,Standard,Monthly,396.0,29.0
4,23.0,Male,32.0,20.0,5.0,8.0,Basic,Monthly,617.0,20.0


In [6]:
y = y.map({'Yes': 1, 'No': 0})
y.head()

,Churn
0,NaN
1,NaN
2,NaN
3,NaN
4,NaN


In [7]:
X = pd.get_dummies(X, drop_first=True)
X.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly
0,30.0,39.0,14.0,5.0,18.0,932.0,17.0,False,False,True,False,False
1,65.0,49.0,1.0,10.0,8.0,557.0,6.0,False,False,False,True,False
2,55.0,14.0,4.0,6.0,18.0,185.0,3.0,False,False,False,False,True
3,58.0,38.0,21.0,7.0,7.0,396.0,29.0,True,False,True,True,False
4,23.0,32.0,20.0,5.0,8.0,617.0,20.0,True,False,False,True,False


In [8]:
X.shape

(440833, 12)

In [9]:
y.isnull().sum()

np.int64(440833)

In [10]:
y = train['Churn']
y.head()

,Churn
0,1.0
1,1.0
2,1.0
3,1.0
4,1.0


In [11]:
y.unique()

array([ 1.,  0., nan])

In [12]:
train = train.dropna(subset=['Churn'])

In [13]:
y = train['Churn']
y.unique()

array([1., 0.])

In [14]:
X = train.drop('Churn', axis=1)
y = train['Churn']

In [15]:
X = X.drop('CustomerID', axis=1)

In [16]:
X = pd.get_dummies(X, drop_first=True)
X.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly
0,30.0,39.0,14.0,5.0,18.0,932.0,17.0,False,False,True,False,False
1,65.0,49.0,1.0,10.0,8.0,557.0,6.0,False,False,False,True,False
2,55.0,14.0,4.0,6.0,18.0,185.0,3.0,False,False,False,False,True
3,58.0,38.0,21.0,7.0,7.0,396.0,29.0,True,False,True,True,False
4,23.0,32.0,20.0,5.0,8.0,617.0,20.0,True,False,False,True,False


In [17]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [18]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

In [19]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=2000)
model.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=2000)

In [21]:
y_pred = model.predict(X_val)

In [22]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_val, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8963898057096192


In [23]:
results = X_val.copy()
results['Actual'] = y_val
results['Predicted'] = y_pred

results.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly,Actual,Predicted
159565,23.0,10.0,16.0,8.0,16.0,589.00,13.0,False,True,False,False,True,1.0,1.0
127464,33.0,3.0,20.0,0.0,22.0,158.00,12.0,True,False,False,False,True,1.0,1.0
22398,27.0,57.0,3.0,0.0,24.0,915.00,26.0,False,False,True,False,False,1.0,0.0
99337,35.0,28.0,23.0,5.0,13.0,334.00,29.0,False,False,False,False,False,1.0,1.0
300997,28.0,53.0,30.0,2.0,5.0,568.86,15.0,True,False,False,False,False,0.0,0.0


In [24]:
churn_risk = results[results['Predicted'] == 1]

churn_risk.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly,Actual,Predicted
159565,23.0,10.0,16.0,8.0,16.0,589.0,13.0,False,True,False,False,True,1.0,1.0
127464,33.0,3.0,20.0,0.0,22.0,158.0,12.0,True,False,False,False,True,1.0,1.0
99337,35.0,28.0,23.0,5.0,13.0,334.0,29.0,False,False,False,False,False,1.0,1.0
55619,41.0,20.0,22.0,0.0,15.0,814.0,6.0,False,False,False,True,False,1.0,1.0
82320,24.0,21.0,2.0,3.0,4.0,542.0,29.0,False,True,False,True,False,1.0,1.0


In [25]:
len(churn_risk)

47981

In [26]:
churn_risk.describe()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Actual,Predicted
count,47981.000000,47981.000000,47981.000000,47981.000000,47981.000000,47981.000000,47981.000000,47981.000000,47981.0
mean,42.150518,30.271587,15.417936,5.389404,15.530835,527.018939,15.746358,0.926929,1.0
std,13.435797,17.364666,8.643131,2.979548,8.672903,248.726950,8.643897,0.260255,0.0
min,18.000000,1.000000,1.000000,0.000000,0.000000,100.000000,1.000000,0.000000,1.0
25%,31.000000,15.000000,8.000000,3.000000,8.000000,313.620000,8.000000,1.000000,1.0
50%,43.000000,30.000000,15.000000,5.000000,16.000000,528.000000,16.000000,1.000000,1.0
75%,53.000000,45.000000,23.000000,8.000000,23.000000,726.710000,23.000000,1.000000,1.0
max,65.000000,60.000000,30.000000,10.000000,30.000000,1000.000000,30.000000,1.000000,1.0


In [27]:
model.predict_proba(X_val)

array([[0.00992814, 0.99007186],
       [0.23648417, 0.76351583],
       [0.8441746 , 0.1558254 ],
       ...,
       [0.94485041, 0.05514959],
       [0.02739154, 0.97260846],
       [0.9795495 , 0.0204505 ]])

In [28]:
proba = model.predict_proba(X_val)

results['Churn_Probability'] = proba[:,1]

results.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly,Actual,Predicted,Churn_Probability
159565,23.0,10.0,16.0,8.0,16.0,589.00,13.0,False,True,False,False,True,1.0,1.0,0.990072
127464,33.0,3.0,20.0,0.0,22.0,158.00,12.0,True,False,False,False,True,1.0,1.0,0.763516
22398,27.0,57.0,3.0,0.0,24.0,915.00,26.0,False,False,True,False,False,1.0,0.0,0.155825
99337,35.0,28.0,23.0,5.0,13.0,334.00,29.0,False,False,False,False,False,1.0,1.0,0.992138
300997,28.0,53.0,30.0,2.0,5.0,568.86,15.0,True,False,False,False,False,0.0,0.0,0.094102


In [29]:
high_risk = results[results['Churn_Probability'] > 0.8]

high_risk.head()

,Age,Tenure,Usage Frequency,Support Calls,Payment Delay,Total Spend,Last Interaction,Gender_Male,Subscription Type_Premium,Subscription Type_Standard,Contract Length_Monthly,Contract Length_Quarterly,Actual,Predicted,Churn_Probability
159565,23.0,10.0,16.0,8.0,16.0,589.0,13.0,False,True,False,False,True,1.0,1.0,0.990072
99337,35.0,28.0,23.0,5.0,13.0,334.0,29.0,False,False,False,False,False,1.0,1.0,0.992138
55619,41.0,20.0,22.0,0.0,15.0,814.0,6.0,False,False,False,True,False,1.0,1.0,0.999408
82320,24.0,21.0,2.0,3.0,4.0,542.0,29.0,False,True,False,True,False,1.0,1.0,0.999984
32582,24.0,10.0,6.0,9.0,25.0,167.0,12.0,True,True,False,False,False,1.0,1.0,0.999624


In [30]:
len(high_risk)

40312

## Model Performance

* Logistic Regression was used for classification.
* The model was trained on historical customer data.
* The model achieved an accuracy of approximately 89.6%, demonstrating strong predictive capability.
* This indicates that the model can reliably predict customer churn.

---

## Prediction Results

The model identified approximately 47,981 customers as likely to churn.

Among them, 40,312 customers were classified as very high risk, with a churn probability above 80%.

Analysis of these high-risk customers shows that:

* The average age is around 42 years old.
* They have a moderate tenure (around 30 months).
* Their average spending is approximately 527.
* Their usage frequency is moderate, indicating regular but not intensive engagement.

These patterns suggest that churn is not limited to low-engagement users, but also affects moderately active customers with average spending behavior.

This insight is valuable for businesses, as it highlights the importance of monitoring not only low-engagement customers, but also mid-level users who may silently churn.

Companies can use this information to design targeted retention strategies such as personalized offers, improved customer engagement, or loyalty incentives.

---

## Key Insights

* The model achieved strong predictive performance with an accuracy of approximately 89.6%.
* Customer behavior such as usage frequency, tenure, and total spending plays a significant role in churn prediction.
* The model enables businesses to identify customers at risk of leaving and take preventive actions.
* Probability-based predictions allow prioritization of high-risk customers, improving decision-making.
* This type of analysis can support customer retention strategies and improve long-term revenue.

---

## Conclusion

This project demonstrates how machine learning can be applied to predict customer churn.

By analyzing customer behavior and training a logistic regression model, we were able to achieve high prediction accuracy.

The use of probability scores allows businesses to move beyond simple classification and focus on risk-based decision-making.

This approach can be valuable for companies aiming to improve customer retention, optimize marketing strategies, and support data-driven decision-making.
